### DISCUSSION TITLE
Limits of ImportC?

### DISCUSSION THREAD

**Discussion by IngwiePhoenix** [dlang/dmd#14653] [General]
Hello!

I tried to compile a [V](https://github.com/vlang) program with the DMD compiler to see what would happen. Well, it actually almost compiled, which I think is pretty awesome!

For context: https://gist.github.com/IngwiePhoenix/d2336f3d2603f87cb0634b79cae7e607

Those few errors asside, what are the actual limitations of `importC` exactly?

In an ideal situation, I'd just "import" my C dependencies from D. The V experiment was basically just "does it work?" calibre of fooling around. But I wondered what I should look out for :)

Thanks and kind regards,
Ingwie

**Comment by ntrel** [comment]
Have you tried compiling the C file with e.g. gcc, do you need to pass any special compiler options?
Which errors do you think are solely due to dmd?

**Reply by IngwiePhoenix** [reply]
I did, with no error. This generated C code is passed immediately to the host's C compiler after generation, its basically the last step of how V "compiles" stuff.

It seems that builtins aren't recognized properly and most of the errors seem to not like the kind of pointer-assignment that V generated from the hashing function (xxHash i believe?). Mainly something like `uint16_t r=*A`.

```
// dmd: test.c(610): Error: cannot implicitly convert expression `*A` of type `ulong` to `__uint128_t`
// dmd: test.c(610): Error: `r` is not a scalar, it is a `__uint128_t`
/* 610 */ __uint128_t r=*A; r*=*B;

// dmd: test.c(614): Error: cannot cast expression `r` of type `__uint128_t` to `ulong`
// dmd: test.c(614): Error: `r` is not of integral type, it is a `__uint128_t`
/* 614 */ *A=(uint64_t)r; *B=(uint64_t)(r>>64);

// dmd: test.c(3286): Error: `[first_bigrams][0]` is not an lvalue and cannot be modified
// dmd: test.c(3286): Error: `[bigram][0]` is not an lvalue and cannot be modified
// dmd: test.c(3286): Error: `[first_bigrams][0]` is not an lvalue and cannot be modified
/* 3286 */ int q = (_IN_MAP(ADDR(string, bigram), ADDR(map, first_bigrams)) ? ((*(int*)map_get(ADDR(map, first_bigrams), &(string[]){bigram}, &(int[]){ 0 })) + 1) : (1));

// dmd: test.c(3292): Error: `[first_bigrams][0]` is not an lvalue and cannot be modified
// dmd: test.c(3292): Error: `[bigram][0]` is not an lvalue and cannot be modified
// dmd: test.c(3292): Error: `[first_bigrams][0]` is not an lvalue and cannot be modified
/* 3292 */ int count = (_IN_MAP(ADDR(string, bigram), ADDR(map, first_bigrams)) ? ((*(int*)map_get(ADDR(map, first_bigrams), &(string[]){bigram}, &(int[]){ 0 }))) : (0));
```

The others are just undefined `__builtin_unreachable` errors, which makes sense as they are built in and probably just not known to DMD. Info: [for GCC](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html) and [for Clang](https://clang.llvm.org/docs/LanguageExtensions.html).
